joomla/joomla-cms
# Expose cache_path config, catch exceptions cleaning cache on config save

### Summary of Changes

1. The code already has support for a `cache_path` parameter which allows a user to specify a custom cache path when using the filesystem for caching.  This parameter is now exposed to the global configuration.

2. When saving the global configuration, the code checking for the filesystem being usable did not account for this parameter and always used the hardcoded `JPATH_SITE . '/cache'` path for this check; the value is now used.  In addition to this, we will also check if a custom cache path was given for this check and cannot be written to but the default path can be, we'll gracefully fall back to the default (and warn the user of this).

3. Clearing the cache store when disabling the cache should not result in a fatal error saving the global configuration.  Wrap the `$cache->clean()` call in try/catch blocks for the `JCacheException*` classes to handle these known errors (note this purposefully does not catch the parent `RuntimeException` as try/catch blocks should generally be specific to known error conditions).

4. Since we clear the cache store when disabling the cache, we should also do it when changing the handler.  That is implemented now.

### Testing Instructions

Saving the global config should still work no matter what.

Setting a value for the `cache_path` var should result in that being the cache path used for the filesystem cache, when empty the code should not set the variable to the saved configuration.  If this path can't be written to, the default path should be used instead and the `cache_path` value not be saved.

When changing cache handlers or disabling an active cache, the cache store should be cleaned.  If the cache store can't be reached then a warning should be shown to the user and not block the save operation.

### Documentation Changes Required

N/A
